Sources/SwiftProtobuf/JSONEncoder.swift

```diff
@@ -148,13 +173,19 @@ internal struct JSONEncoder {
                 append(staticText: "\"Infinity\"")
             }
         } else {
-            // TODO: Be smarter here about choosing significant digits
-            // See: protoc source has C++ code for this with interesting ideas
             if let v = Int64(safely: value) {
                 appendInt(value: v)
             } else {
                 let s = String(value)
-                append(text: s)
+                let reparsed = Double(s)
+                // If exact match, then default precision is sufficient:
+                if value == reparsed {
+                    append(text: s)
+                } else {
+                    let precision = DBL_DIG + 2
+                    let s = String(format: "%.*g", precision, value)
```

Comment:
Good point.

I added a bunch of tests to JSON and Text to verify round-tripping of floating-point numbers through those formats.  These tests are able to reproduce the conformance failures for Double (but not for Float for reasons I don't yet understand).

I've updated Text with the same logic as JSON.

I checked the plugin code as well:  The descriptor that protoc sends to the plugin contains a textual form of the default value.  Apart from translating "inf" and "nan", we just pass the literal text straight through to the Swift source code.  So the numeric literal that gets written to the Swift source code is actually formatted by the C++ library code.